lib/uplink/project.go

```diff
@@ -204,11 +195,40 @@ func (p *Project) OpenBucket(ctx context.Context, bucketName string, access *Enc
 		Name:         bucketInfo.Name,
 		Created:      bucketInfo.Created,
 		bucket:       bucketInfo,
-		metainfo:     kvmetainfo.New(p.project, p.metainfo, streamStore, segmentStore, encStore),
+		metainfo:     kvmetainfo.New(p.project, p.metainfo, streamStore, segmentStore, encCtx.store),
 		streams:      streamStore,
 	}, nil
 }
 
+func (p *Project) retrieveSalt(ctx context.Context) (salt []byte, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	info, err := p.metainfo.GetProjectInfo(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return info.ProjectSalt, nil
+}
+
+// SaltedKeyFromPassphrase returns a key generated from the given passphrase using a stable, project-specific salt
+func (p *Project) SaltedKeyFromPassphrase(ctx context.Context, passphrase string) (_ *storj.Key, err error) {
+	defer mon.Task()(&ctx)(&err)
+	salt, err := p.retrieveSalt(ctx)
+	if err != nil {
+		return nil, err
+	}
+	key, err := encryption.DeriveDefaultPassword([]byte(passphrase), salt)
```

Comment:
Is there a reason derive isn't returning storj.Key?

Reply:
No, that's fine. It's a `[]byte` because it originally wasn't designed to be used as a key, but that changed. I also have to update some design docs.